Mathlib/Analysis/SpecialFunctions/Log/Summable.lean

```diff
@@ -8,47 +8,91 @@ import Mathlib.Analysis.SpecialFunctions.Complex.LogBounds
+variable {ι : Type*}
 
-open scoped Interval Topology BigOperators Nat Complex
+open Filter Topology NNReal
 
-variable {ι : Type*}
+namespace Complex
+variable {f : ι → ℂ} {a : ℂ}
+
+lemma hasProd_of_hasSum_log (hfn : ∀ i, f i ≠ 0) (hf : HasSum (fun i ↦ log (f i)) a) :
+    HasProd f (exp a) :=
+  hf.cexp.congr (by simp [exp_log, hfn])
+
+lemma multipliable_of_summable_log (hf : Summable fun i ↦ log (f i)) :
+    Multipliable f := by
+  by_cases hfn : ∃ n, f n = 0
+  · exact multipliable_of_exists_eq_zero hfn
+  · exact ⟨_, hasProd_of_hasSum_log (not_exists.mp hfn) hf.hasSum⟩
+
+/-- The exponential of a convergent sum of complex logs is the corresponding infinite product. -/
+lemma cexp_tsum_eq_tprod (hfn : ∀ i, f i ≠ 0) (hf : Summable fun i ↦ log (f i)) :
+    cexp (∑' i, log (f i)) = ∏' i, f i :=
+  (hasProd_of_hasSum_log hfn hf.hasSum).tprod_eq.symm
 
-lemma Complex.summable_log_one_add_of_summable {f : ι → ℂ} (hf : Summable f) :
-    Summable (fun i : ι => Complex.log (1 + f i)) := by
-  apply (hf.norm.const_smul (3 / 2 : ℝ)).of_norm_bounded_eventually
+lemma summable_log_one_add_of_summable {f : ι → ℂ} (hf : Summable f) :
+    Summable (fun i ↦ log (1 + f i)) := by
+  apply (hf.norm.mul_left _).of_norm_bounded_eventually
   filter_upwards [hf.norm.tendsto_cofinite_zero.eventually_le_const one_half_pos] with i hi
-  exact norm_log_one_add_half_le_self hi
-
-lemma Real.summable_log_one_add_of_summable {f : ι → ℝ} (hf : Summable f) :
-     Summable (fun i : ι => log (1 + |f i|)) := by
-  have : Summable (fun n ↦ Complex.ofRealCLM (log (1 + |f n|))) := by
-    convert Complex.summable_log_one_add_of_summable (Complex.ofRealCLM.summable hf.norm) with x
-    rw [ofRealCLM_apply, ofReal_log (by positivity)]
-    simp only [ofReal_add, ofReal_one, norm_eq_abs, ofRealCLM_apply]
-  convert Complex.reCLM.summable this
-
-lemma Complex.multipliable_one_add_of_summable (f : ι → ℂ) (hf : Summable f)
-    (hff : ∀ n : ι, 1 + f n ≠ 0) : Multipliable (fun n : ι => 1 + f n) := by
-  refine Complex.multipliable_of_summable_log (fun n => 1 + f n) (by simpa) ?_
-  simpa only [forall_const] using Complex.summable_log_one_add_of_summable hf
-
-lemma Real.multipliable_one_add_of_summable (f : ι → ℝ) (hf : Summable f) :
-    Multipliable (fun n : ι => 1 + |f n|) := by
-  refine Real.multipliable_of_summable_log (fun n => 1 + |f n|) (fun _ ↦ by positivity) ?_
-  simpa only [forall_const] using Real.summable_log_one_add_of_summable  hf
+    using norm_log_one_add_half_le_self hi
+
+lemma multipliable_one_add_of_summable (hf : Summable f) : Multipliable (fun i ↦ 1 + f i) :=
+  multipliable_of_summable_log (summable_log_one_add_of_summable hf)
+
+end Complex
+
+namespace Real
+variable {f : ι → ℝ} {a : ℝ}
+
+lemma hasProd_of_hasSum_log (hfn : ∀ i, 0 < f i) (hf : HasSum (fun i ↦ log (f i)) a) :
+    HasProd f (rexp a) :=
+  hf.rexp.congr (by simp [exp_log, hfn])
+
+lemma multipliable_of_summable_log (hfn : ∀ i, 0 < f i) (hf : Summable fun i ↦ log (f i)) :
```

Comment:
Again, `hfn` is probably not needed here, and the next lemma can be removed.

Reply:
No, this one’s needed, because `Real.exp (-1) =0`, so you could have a sequence randomly oscillating in sign and then its log would be dimmable but it wouldn’t be multipliable.